### PR TITLE
Add DNS provider for SelfHost.(de|eu)

### LINKED ISSUE #2277: Support for selfhost
### Welcome

- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, the DNS provider exposes a public API.
- [X] Yes, I know that the lego maintainers don't have an account in all DNS providers in the world.
- [ ] Yes, I'm able to create a pull request and be able to maintain the implementation.
- [X] Yes, I'm able to test an implementation if someone creates a pull request to add the support of this DNS provider.

### How do you use lego?

Through Traefik

### Link to the DNS provider

https://www.selfhost.de

### Link to the API documentation

?

### Additional Notes

There is an implementation inside `acme.sh`: https://github.com/acmesh-official/acme.sh/blob/master/dnsapi/dns_selfhost.sh

https://github.com/traefik/traefik/issues/11096

### PR DESCRIPTION
**Pull request by ldez**
- [x] adds a description to your PR
- [x] have a homogeneous design with the other providers
- [x] add tests (units)
- [x] add tests ("live")
- [x] add a [provider descriptor](https://github.com/go-acme/lego/blob/master/providers/dns/cloudflare/cloudflare.toml)
- [x] generate CLI help, documentation, and readme.
- [x] be able to do: _(and put the output of this command to a comment)_
  ```bash
  make build
  rm -rf .lego

  SELFHOSTDE_USERNAME=xxx \
  SELFHOSTDE_PASSWORD=yyy \
  SELFHOSTDE_RECORDS_MAPPING=example.com:123:456 \
  ./dist/lego -m your@email.com --dns selfhostde -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
  ```
  Note the wildcard domain is important.
- [x] pass the linter
- [x] do `go mod tidy`

Ping @aifrog, can you run the command (with your domain, email, credentials, etc.)?

-> https://github.com/go-acme/lego/pull/2278#issuecomment-2362213145

Closes #2277

### CODE REVIEW

**Comment by ldez** [comment]
the mapping should not contain `_acme-challenge.`

```
SELFHOSTDE_RECORDS_MAPPING=pve1.aaa.bbb:1234:1234 
```

**Comment by aifrog** [comment]
An maybe noteworthy: _acme-challenge.pve1.aaa.bbb txt entry exists in the DNS

**Comment by ldez** [comment]
The problem is not the TXT record but the mapping you provided: the mapping should not contain `_acme-challenge.` prefix.

I added a fallback to handle this case but you don't need to provide this prefix.

**Comment by ldez** [comment]
You should create a TXT record with this name: `_acme-challenge.pve1.aaa.bbb`
But inside the mapping (`SELFHOSTDE_RECORDS_MAPPING`), you need to use `pve1.aaa.bbb`.

I don't forget to pull the latest commit of my PR.

**Comment by aifrog** [comment]
I now see that it fills in a string in the DNS and waits. After a while it fails and enters "empty". The error at the end is:

**Comment by ldez** [comment]
I think you forget to add the logs

**Comment by aifrog** [comment]
I can see that a string is entered in the DNS successfully. It then waits for the propagation and this fails after a while:

2024/09/18 17:23:11 [INFO] [aaa.bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/18 17:23:12 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14058553123
2024/09/18 17:23:12 Could not obtain certificates:
        error: one or more domains had a problem:
[aaa.bbb.ccc] propagation: time limit exceeded: last error: NS ns.bbb.ccc. returned NXDOMAIN for _acme-challenge.aaa.bbb.ccc.

Last, I see that the DNS entry is changed to "empty".

**Comment by ldez** [comment]
> Last, I see that the DNS entry is changed to "empty".

This is expected: as we cannot delete a TXT record, I just put `empty` as a value.

> It then waits for the propagation and this fails after a while

Can you try to increase the propagation timeout:

```bash
SELFHOSTDE_USERNAME=xxx \
SELFHOSTDE_PASSWORD=yyy \
SELFHOSTDE_RECORDS_MAPPING=example.com:123:456 \
SELFHOSTDE_PROPAGATION_TIMEOUT=240 \
SELFHOSTDE_POLLING_INTERVAL=30 \
./dist/lego -m your@email.com --dns selfhostde -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

**Comment by aifrog** [comment]
It does not work but I might have found the problem:
The DNS entry starts with _acme_challenge and gets filled in correctly.
The error message mentions _acme-challenge which is (slightly) different
Why not change both to _acme-challenge?

**Comment by aifrog** [comment]
It now works :-) Thank you!!

**Comment by ldez** [comment]
can you give me the full output (with your personal information redacted)?

I need to see the `rm -rf .lego` and all the logs.

**Comment by aifrog** [comment]
I don't understand the "I need to see the rm -rf .lego and all the logs." but here is the output:

SELFHOSTDE_USERNAME=1234 SELFHOSTDE_PASSWORD=1234 SELFHOSTDE_RECORDS_MAPPING=aaa.bbb.ccc:1234:1234 SELFHOSTDE_PROPAGATION_TIMEOUT=240 SELFHOSTDE_POLLING_INTERVAL=30 ./dist/lego -m xxx@bbb.ccc --dns.resolvers 1.1.1.1 --dns selfhostde -d aaa.bbb.ccc -s https://acme-staging-v02.api.letsencrypt.org/directory run
2024/09/18 20:41:23 [INFO] [aaa.bbb.ccc] acme: Obtaining bundled SAN certificate
2024/09/18 20:41:24 [INFO] [aaa.bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14059388453
2024/09/18 20:41:24 [INFO] [aaa.bbb.ccc] acme: authorization already valid; skipping challenge
2024/09/18 20:41:24 [INFO] [aaa.bbb.ccc] acme: Validations succeeded; requesting certificates
2024/09/18 20:41:24 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2024/09/18 20:41:27 [INFO] [aaa.bbb.ccc] Server responded with a certificate.

**Comment by aifrog** [comment]
SELFHOSTDE_USERNAME=1234 SELFHOSTDE_PASSWORD=1234 SELFHOSTDE_RECORDS_MAPPING=aaa.bbb.ccc:1234:1234 SELFHOSTDE_PROPAGATION_TIMEOUT=240 SELFHOSTDE_POLLING_INTERVAL=30 ./dist/lego -m xxx@bbb.ccc --dns.resolvers 1.1.1.1 --dns selfhostde -d aaa.bbb.ccc -s https://acme-staging-v02.api.letsencrypt.org/directory run
2024/09/18 20:51:20 No key found for account xxx@bbb.ccc. Generating a P256 key.
2024/09/18 20:51:20 Saved key to /home/xxx/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/xxx@bbb.ccc/keys/xxx@bbb.ccc.key
2024/09/18 20:51:20 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.4-April-3-2024.pdf
Do you accept the TOS? Y/n
Y
2024/09/18 20:51:26 [INFO] acme: Registering account for xxx@bbb.ccc
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/xxx/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] acme: Obtaining bundled SAN certificate
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14060601223
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] acme: Could not find solver for: tls-alpn-01
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] acme: Could not find solver for: http-01
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] acme: use dns-01 solver
2024/09/18 20:51:27 [INFO] [aaa.bbb.ccc] acme: Preparing to solve DNS-01
2024/09/18 20:51:28 [INFO] [aaa.bbb.ccc] acme: Trying to solve DNS-01
2024/09/18 20:51:28 [INFO] [aaa.bbb.ccc] acme: Checking DNS record propagation. [nameservers=1.1.1.1:53]
2024/09/18 20:51:58 [INFO] Wait for propagation [timeout: 4m0s, interval: 30s]
2024/09/18 20:52:06 [INFO] [aaa.bbb.ccc] The server validated our request
2024/09/18 20:52:06 [INFO] [aaa.bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/18 20:52:06 [INFO] [aaa.bbb.ccc] acme: Validations succeeded; requesting certificates
2024/09/18 20:52:06 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2024/09/18 20:52:08 [INFO] [aaa.bbb.ccc] Server responded with a certificate.

**Comment by ldez** [comment]
You are not executing the required commands.

```
rm -rf .lego

SELFHOSTDE_USERNAME=xxx \
SELFHOSTDE_PASSWORD=yyy \
SELFHOSTDE_RECORDS_MAPPING=example.com:123:456 \
SELFHOSTDE_PROPAGATION_TIMEOUT=240 \
SELFHOSTDE_POLLING_INTERVAL=30 \
./dist/lego -m your@email.com --dns selfhostde -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

Note there are 2 domains: a wildcard domain (`*.example.com`) and the same domain without wildcard (`example.com`).

**Comment by aifrog** [comment]
Too bad, this just stops and never returns:

SELFHOSTDE_USERNAME=1234 SELFHOSTDE_PASSWORD=1234 SELFHOSTDE_RECORDS_MAPPING=aaa.bbb.ccc:1234:1234 SELFHOSTDE_PROPAGATION_TIMEOUT=240 SELFHOSTDE_POLLING_INTERVAL=30 ./dist/lego -m xxx@bbb.ccc --dns.resolvers 1.1.1.1 --dns selfhostd
e -d *.bbb.ccc -d bbb.ccc -s https://acme-staging-v02.api.letsencrypt.org/directory run
2024/09/18 21:23:15 [INFO] [*.bbb.ccc, bbb.ccc] acme: Obtaining bundled SAN certificate
2024/09/18 21:23:16 [INFO] [*.bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14060847013
2024/09/18 21:23:16 [INFO] [bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14060885213
2024/09/18 21:23:16 [INFO] [bbb.ccc] acme: Could not find solver for: tls-alpn-01
2024/09/18 21:23:16 [INFO] [bbb.ccc] acme: Could not find solver for: http-01
2024/09/18 21:23:16 [INFO] [bbb.ccc] acme: use dns-01 solver
2024/09/18 21:23:16 [INFO] [*.bbb.ccc] acme: use dns-01 solver
2024/09/18 21:23:16 [INFO] [bbb.ccc] acme: Preparing to solve DNS-01
2024/09/18 21:23:16 [INFO] [*.bbb.ccc] acme: Preparing to solve DNS-01

**Comment by ldez** [comment]
I updated the PR, can you pull the changes and retry?

**Comment by aifrog** [comment]
2024/09/18 22:19:26 [INFO] [*.bbb.ccc] acme: Preparing to solve DNS-01
2024/09/18 22:19:26 [INFO] [bbb.ccc] acme: Preparing to solve DNS-01
2024/09/18 22:19:26 [INFO] [*.bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/18 22:19:26 [WARN] [*.bbb.ccc] acme: cleaning up failed: selfhostde: unknown record ID for 'bbb.ccc.'
2024/09/18 22:19:26 [INFO] [bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/18 22:19:26 [WARN] [bbb.ccc] acme: cleaning up failed: selfhostde: unknown record ID for 'bbb.ccc.'
2024/09/18 22:19:26 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14061401983
2024/09/18 22:19:26 [INFO] Deactivating auth: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14061401993
2024/09/18 22:19:27 Could not obtain certificates:
        error: one or more domains had a problem:
[*.bbb.ccc] [*.bbb.ccc] acme: error presenting token: selfhostde: record mapping not found for bbb.ccc
[bbb.ccc] [bbb.ccc] acme: error presenting token: selfhostde: record mapping not found for bbb.ccc

**Comment by ldez** [comment]
are you providing the right mapping?

From this comment: https://github.com/go-acme/lego/pull/2278#issuecomment-2359245103

You defined a mapping for `aaa.bbb.ccc` but you are trying to get certificates for `bbb.ccc`

**Comment by aifrog** [comment]
Still not working. I provided _acme-challenge.bbb.ccc and I see that there is change in the txt field. However, it is not filled with "empty" as before. I do not know if I need another entry for "*.bbb.ccc"

**Comment by ldez** [comment]
can you provide the logs?

**Comment by ldez** [comment]
You need to create 2 TXT records for the same domain (`_acme-challenge.example.com`) to have 2 different IDs.

```bash
SELFHOSTDE_USERNAME=xxx \
SELFHOSTDE_PASSWORD=yyy \
SELFHOSTDE_RECORDS_MAPPING=example.com:<RECORD_ID1>:<RECORD_ID2>  \
./dist/lego -m xxx@example.com --dns.resolvers 1.1.1.1 --dns selfhostde -d *.example.com -d example.com -s https://acme-staging-v02.api.letsencrypt.org/directory run
```

**Comment by aifrog** [comment]
Thats it. Works:

SELFHOSTDE_USERNAME=1234 SELFHOSTDE_PASSWORD=1234 SELFHOSTDE_RECORDS_MAPPING=bbb.ccc:1234:1234 SELFHOSTDE_PROPAGATION_TIMEOUT=240 SELFHOSTDE_POLLING_INTERVAL=30 ./dist/lego -m xxx@bbb.ccc --dns.resolvers 1.1.1.1 --dns selfhostde -d *.bbb.ccc -d bbb.ccc -s https://acme-staging-v02.api.letsencrypt.org/directory run
2024/09/19 23:10:30 No key found for account xxx@bbb.ccc. Generating a P256 key.
2024/09/19 23:10:30 Saved key to /home/xxx/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/xxx@bbb.ccc/keys/xxx@bbb.ccc.key
2024/09/19 23:10:31 Please review the TOS at https://letsencrypt.org/documents/LE-SA-v1.4-April-3-2024.pdf
Do you accept the TOS? Y/n
Y
2024/09/19 23:10:35 [INFO] acme: Registering account for xxx@bbb.ccc
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/xxx/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2024/09/19 23:10:35 [INFO] [*.bbb.ccc, bbb.ccc] acme: Obtaining bundled SAN certificate
2024/09/19 23:10:36 [INFO] [*.bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14075870753
2024/09/19 23:10:36 [INFO] [bbb.ccc] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/14075870763
2024/09/19 23:10:36 [INFO] [*.bbb.ccc] acme: use dns-01 solver
2024/09/19 23:10:36 [INFO] [bbb.ccc] acme: Could not find solver for: tls-alpn-01
2024/09/19 23:10:36 [INFO] [bbb.ccc] acme: Could not find solver for: http-01
2024/09/19 23:10:36 [INFO] [bbb.ccc] acme: use dns-01 solver
2024/09/19 23:10:36 [INFO] [*.bbb.ccc] acme: Preparing to solve DNS-01
2024/09/19 23:10:51 [INFO] [bbb.ccc] acme: Preparing to solve DNS-01
2024/09/19 23:10:53 [INFO] [*.bbb.ccc] acme: Trying to solve DNS-01
2024/09/19 23:11:23 [INFO] Wait for propagation [timeout: 4m0s, interval: 30s]n. [nameservers=1.1.1.1:53]
2024/09/19 23:11:31 [INFO] [*.bbb.ccc] The server validated our request
2024/09/19 23:11:31 [INFO] [bbb.ccc] acme: Trying to solve DNS-01
2024/09/19 23:11:31 [INFO] [bbb.ccc] acme: Checking DNS record propagation. [nameservers=1.1.1.1:53]
2024/09/19 23:12:01 [INFO] Wait for propagation [timeout: 4m0s, interval: 30s]
2024/09/19 23:12:07 [INFO] [bbb.ccc] The server validated our request
2024/09/19 23:12:07 [INFO] [*.bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/19 23:12:08 [INFO] [bbb.ccc] acme: Cleaning DNS-01 challenge
2024/09/19 23:12:08 [INFO] [*.bbb.ccc, bbb.ccc] acme: Validations succeeded; requesting certificates
2024/09/19 23:12:08 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2024/09/19 23:12:10 [INFO] [*.bbb.ccc] Server responded with a certificate.

File: providers/dns/selfhostde/internal/readme.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# SelfHost.(de|eu)
+
+SelfHost doesn't provide an official API documentation and there are no endpoints for create a TXT record or delete a TXT record.
+
+## More
+
+This link (https://kirk.selfhost.de/cgi-bin/selfhost?p=document&name=api) content a PDF that doesn't describe the endpoint used by the client.
```
Comment (dmke) [inline providers/dns/selfhostde/internal/readme.md:7]:
```suggestion
The documentation found at https://kirk.selfhost.de/cgi-bin/selfhost?p=document&name=api (PDF) describes the DynDNS/ddns API endpoint and is not used by our client.
```

File: providers/dns/selfhostde/selfhostde.toml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+SELFHOSTDE_PASSWORD=yyy \
+SELFHOSTDE_RECORDS_MAPPING=my.example.com:123 \
+lego --email you@example.com --dns selfhostde --domains my.example.org run
+'''
+
+Additional = """
+SelfHost.de doesn't have an API to create or delete TXT records, there is only an "unofficial" and undocumented endpoint to update an existing TXT record.
+
+So, before using lego to request a certificate for a given domain or wildcard (such as `my.example.org` or `*.my.example.org`),
+you should create:
```
Comment (dmke) [inline providers/dns/selfhostde/selfhostde.toml:18]:
```suggestion
you must create:
```

File: providers/dns/selfhostde/selfhostde.toml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+You should recreate a mapping to fill the `SELFHOSTDE_RECORDS_MAPPING` environement variable as following:
+
+```
+<domain_A>:<record_id_A1>:<record_id_A2>,<domain_B>:<record_id_B1>:<record_id_B2>,<domain_C>:<record_id_C1>:<record_id_C2>
+```
+
+Each group of domain + record id(s) is separated with a comma `,`.
+
+Each record id is separated with 2 dots `:`.
```
Comment (dmke) [inline providers/dns/selfhostde/selfhostde.toml:32]:
````suggestion
You then must prepare the `SELFHOSTDE_RECORDS_MAPPING` environment variable with the following format:

```
<domain_A>:<record_id_A1>:<record_id_A2>,<domain_B>:<record_id_B1>:<record_id_B2>,<domain_C>:<record_id_C1>:<record_id_C2>
```

where each group of domain + record ID(s) is separated with a comma (`,`), and the domain and record ID(s) are separated with a colon (`:`).

For example, if you want to create or renew a certificate for `my.example.org`, `*.my.example.org`, and `other.example.rog`, you would need:

- two separate records for `_acme-challenge.my.example.org`, and
- another separate record for `_acme-challenge.other.example.org`.

The resulting environment variable would then be: `SELFHOSTDE_RECORDS_MAPPING=my.example.com:123:456,other.example.com:789`

````

**Review by dmke — CHANGES_REQUESTED** [review]
I've reworded the documentation a bit. Let me know if you disagree or have further notes. Otherwise this looks OK. :+1:

**Review by dmke — APPROVED** [review]

**Comment by aifrog** [comment]
Just a short question: I want to use this in traefik and doe have v3.2.0-rc1 running. However, i still get the error="cannot get ACME client unrecognized DNS provider: selfhostde". Is this a configuration issue or is this code not yet part of  v3.2.0-rc1?

**Comment by aifrog** [comment]
I understand, that younhabe released by now. Correct? But I can't see it yet in traefik (v3.2.0-rc2). Do I need to wait or isbit my mistake?

**Comment by ldez** [comment]
lego has been updated inside Traefik, but you must wait for a Traefik release.

**Pull request merged**